element_array_ephys/export/nwb/nwb.py

```diff
@@ -0,0 +1,600 @@
+
+    Returns
+    -------
+    dict
+
+    """
+    return {
+        (electrodes["group"][idx].device.name, electrodes["id_in_probe"][idx],): idx
+        for idx in range(len(electrodes))
+    }
+
+
+def gains_helper(gains):
+    if all(x == 1 for x in gains):
+        return dict(conversion=1e-6, channel_conversion=None)
+    if all(x == gains[0] for x in gains):
+        return dict(conversion=1e-6 * gains[0], channel_conversion=None)
+    return dict(conversion=1e-6, channel_conversion=gains)
+
+
+def add_ephys_recording_to_nwb(
+    session_key: dict, ephys_root_data_dir,
+    nwbfile: pynwb.NWBFile, end_frame: int = None):
+    """Read voltage data directly from source files and iteratively transfer them to the NWB file. Automatically
+    applies lossless compression to the data, to the final file might be smaller than the original, but there is no
+    data loss. Currently supports neuropixel and openephys, and relies on SpikeInterface to read the data.
+
+    source data -> acquisition["ElectricalSeries"]
+
+    Parameters
+    ----------
+    session_key: dict
+    nwbfile: NWBFile
+    end_frame: int, optional
+        Used for small test conversions
+    """
+
+    if nwbfile.electrodes is None:
+        add_electrodes_to_nwb(session_key, nwbfile)
+
+    mapping = get_electrodes_mapping(nwbfile.electrodes)
+
+    for ephys_recording_record in (ephys.EphysRecording & session_key).fetch(
+        as_dict=True
+    ):
+        probe_id = (ephys.ProbeInsertion() & ephys_recording_record).fetch1("probe")
+
+        relative_path = (
+            ephys.EphysRecording.EphysFile & ephys_recording_record
+        ).fetch1("file_path")
+        relative_path = relative_path.replace("\\","/")
+        file_path = find_full_path(ephys_root_data_dir, relative_path)
+
+        if ephys_recording_record["acq_software"] == "SpikeGLX":
+            extractor = extractors.read_spikeglx(os.path.split(file_path)[0], "imec.ap")
+        elif ephys_recording_record["acq_software"] == "OpenEphys":
+            extractor = extractors.read_openephys(file_path, stream_id="0")
+        else:
+            raise ValueError(
+                f"unsupported acq_software type: {ephys_recording_record['acq_software']}"
+            )
+
+        conversion_kwargs = gains_helper(extractor.get_channel_gains())
+
+        if end_frame is not None:
+            extractor = extractor.frame_slice(0, end_frame)
+
+        recording_channels_by_id = (
+            probe.ElectrodeConfig.Electrode() & ephys_recording_record
+        ).fetch("electrode")
+
+        nwbfile.add_acquisition(
+            pynwb.ecephys.ElectricalSeries(
+                name=f"ElectricalSeries{ephys_recording_record['insertion_number']}",
+                description=str(ephys_recording_record),
+                data=SpikeInterfaceRecordingDataChunkIterator(extractor),
+                rate=ephys_recording_record["sampling_rate"],
+                starting_time=(
```

Comment:
I thought the `starting_time` is a timestamp ([reference](https://pynwb.readthedocs.io/en/stable/pynwb.ecephys.html#pynwb.ecephys.ElectricalSeries))?

Reply:
@kabilar All `TimeSeries` can be define sample times either by (starting_time & rate) or timestamps. If the timestamps are regularly sampled, it is best practice to use starting_time and rate. It is true that even a series is supposed to be regularly sampled there may be clock slippage, but the electrophysiology acquisition system is usually the master clock, so I wouldn't expect clock slippage for this `ElectricalSeries`, so I chose to use starting_time and rate here instead of timestamps. Either way, the starting_time or the first value of timestamps should be the starting time of the recording in seconds with respect to the timestamp_reference_time if it is defined, and if not then with respect to the session_start_time. The issue here is that the first value of the LFP timestamps was 0, so I am not sure if that is wrt the start of the LFP series or the start of the session. It seemed like the safest thing was to use this value, but feel free to change this to using timestamps or to use a different start time if I didn't interpret the DJ schema correctly.

Reply:
Thanks for clarifying @bendichter!